### ISSUE TITLE
Record: additional metrics, clustering and filtering

### ISSUE THREAD

**Issue by guancio** [Runalyze/Runalyze#1688]
Three new features for the Record plug-in:
1. Additional metrics e.g. elevation gains, calories, vdot, trimp, ...

2. Filtering: use something similar to the "SEARCH FOR ACTIVITIES" page to limit the activities selected (e.g. only activities longer than 20km, only on the route xxx, only with a elevation gain of at least 800 m)

3. Clustering: records are currently clustered by sport. Someone can be interested in clustering by additional metrics: e.g. Running and distance >= 42km, Running and distance >= 20 km, Running and distance >= 10 km ...

**Comment by mipapo** [comment]
Filtering/searching should be a general function for statistic plugins

**Comment by guancio** [comment]
I'm experimenting with a record plug in any my idea is to define sport specific clusters (like what there is already for the plug-in race and sport running).
The idea is that for running one can be interested in records for 1, 3, 5, 10, 21, 42 kilometres, while for biking it can be 10, 20, 50, 100, 200 and for swimming 0.1, 0.2, 0.5, 0.8, 1, 2, 5 etc.
The plug-in should enable to infer a default value for each sport and should allow the user to change such numbers.

Records are computed using with the activities that satisfy the threshold (e.g. distance > 5 km). Thus an activity having distance 7 km is used in computing the record for 1, 3 and 5 km but not for 10, 21 and 42.

I imagine that for some sports the distance clustering is naive. For mountaineering records should be clustered by elevation gain. For some other sports (e.g. biking) records should be clustered by more than one measure (e.g. distance > 40km and mean gradient > 2%, distance > 80km and mean gradient > 1%, etc).

Once we have record cluster for sport we can enrich the activity view with a record section. Let assume we have a running activity of 7 km in December 9th 2015. The new section can present (for each compatible running cluster: 1, 3, 5 kilometres) the position of the activity in the record chart, together with the 5 nearest other performances. The user should be able to select if the chart has to consider activities of all years, only activity of 2015, only activity of December 2015, only activities up to December 9th 2015.

Additionally, we can create a "virtual cluster" based on the current activity: e.g. a record chart containing all activities that are similar to the current one (i.e. distance +- 10%, gradient +- 5%, etc).

This also enable a further section for the activity view: "achievements". This can be similar to the Strava "TOP RESULTS/achievements". This can also be used to generate motivational alert: if you add an activity that has at least an achievement, then a special box is shown on the home page :).

**Comment by laufhannes** [comment]
That sounds really great - but quite challenging as well.

Assuming we have a defined set of records of interest for a given sports type (see related #1479 - currently running is as such identifiable), we have to make sure to efficiently process and save these achievements. For every activity of more than 5k the best segments for 1k/3k/5k have to be saved somehow.

Keep us up to date on your progess!

**Comment by laufhannes** [comment]
@mipapo just reminded me that @grzeg1 already did some work on records/best segments some time ago: https://github.com/grzeg1/runalyze-test/commits/master

**Comment by mipapo** [comment]
@guancio We just had a look into your branch `new_record`. It's great that you are working on new features, but I just tested some queries against our productive database and their performance is quite bad. That may be negligible for small installations, but it's not for our official installation.
> We just wanted to inform you before you put too much work into it. This will definitely need some 'planning' before it's 'online-ready'.
E.g. the calculation of Records for a individual has to be saved in the database and shouldn't be recalculated everytime.

**Comment by guancio** [comment]
@mipapo thanks for alerting me. I've experimented a little bit just to understand the underlying platform. I never written Php before and is something like 6 years that I do not use a SQL database.
My main goal is to have a base where I can experiment the features that I would like and how these can require complex changes in the architecture.

**Comment by guancio** [comment]
Hi all,
I'm planning to write a small task plan for this task, so that you can review my idea before I continue the development.
Regarding the performance (even I agree that the records should be stored in a table) I think that we should add a pace field to activity. Lot of queries sort by (distance/time) and since these are computed fields, these sorts require probably a table scan (I do not know if MySQL is smart enough to enable a B-tree index on a computed field).

**Comment by mipapo** [comment]
I don't think a pace field is needed. The calculation should be occur in a background running task - It shouldn't be visible to the user. But for that we finally need a worker/cronjob functionality which hasn't been implemented for now

**Comment by guancio** [comment]
My point for adding the field pace is to enable a sorting index (I agree that it is probably not needed for offline computations). In the current schema there are not indexes, can I ask why?

**Comment by mipapo** [comment]
https://github.com/Runalyze/Runalyze/blob/master/inc/install/structure.sql#L491 ?

**Comment by guancio** [comment]
Sorry, my MySQL is really bad :) so ADD KEY = ADD INDEX :-)

**Comment by guancio** [comment]
The following is my idea for the new record functionality.

Records should work per cluster, not per sport. We call a cluster a set of activities meeting a predefined set of requirements. For now I assume that a cluster can include all activities that have
- a specific sport_id
- a distance greater that the cluster threshold.

Some examples: for running I expect the standard clusters to be distance >= 1.0 km, distance >= 5.0 km, distance >= 10.0 km, distance >= 21 km and distance >= 42 km. 

We need a new table to keep track of this configuration: “clusters”. The table should have the following columns:
- id: unique identifier
- sport_id: I assume that it is pointless to create sport independent clusters
- others: the data necessary to identify a cluster. Here we have several options.
    + others is only a column, representing a distance threshold: an activity belongs to the cluster if its distance is greater or equal to the threshold
    + others consists of triples of operator, field and value; if we want the set of all activities whose distance is at least 10 km and and whose altitude is at most 500 m we have op1 = '>=', field1 = 'distance', value1 = '10', op2 = '<=', field2 = 'altitude', value2 = '500', op3 = 'null'
    + others is the json serialization of some php data-structure, used to implement the filter

In the configuration view, we have a new tab to configure clusters. Probably, we can just list all clusters grouped by sports, allowing the user to delete a cluster, change the cluster attributes an add a new cluster to one of the sports.

The clusters table should be updated whenever the sport lists is changed:
- all clusters belonging to a sport are dropped if the sport is deleted
- whenever a new sport is created, the corrsponding standard clusters are created. For the running sport we have 1 km, 2km, 5km, 10km, 21km and 42 km. For the other sports we create a standard configuration: 0km,  5km, 10km, 20km, 50km. In future, we can refine this mechanism to take into account sports like swimming, cycling, etc.

To efficiently match activities with clusters we keep a index table (the reason for such table is explained later) whose name is cluster_training and containing the following column:
- cluster_id
- tranining_id
- a column for every possible field such that a record can be computed (e.g. pace, vdot)

The cluster_training table has a BTree index for every (cluster_id, record_field).

The reason for such index table is the following. Let assume that we want the top ten (by pace) running activities of at least 10 km. This is currently computed as
 ```
SELECT *
FROM tranning
WHERE sport_id=:running and distance>=10km
ORDER by (ditance/s)
LIMIT 10
```
This query is terribly slow, since it is ordered on a computed field, thus requiring a complete tablescan. Even if we assume that the training table has a precomputed column pace, and an index on (sport_id, pace) the query will be still slow. The BTree will sort all running activities by pace, however the DB will have to scan the index and filter out the activities having distance smaller that 10km (btw these are probably the best activity by peace).

On the other hand, with the cluster mechanism, the same task can be solved by
```
SELECT *
FROM cluster_training JOIN tranning
  ON cluster_training.training_id=tranning.id
WHERE cluster_training.id = :cluster_id
ORDER by cluster_training.pace
LIMIT 10
```
This query will be probably be fast, since does not require a table scan. Additionally it has to loop up only the first ten entries of the index. 

**Record-statistics.**
A plug-in allows the user to see his ten best activities by cluster. This task can be easily accomplished by using the query above. We can also provide a paginated version , where the user can iterate among his sorted activities and see ten activities per page.
An additional functionality can be display the records by time range (e.g. all, this year, last year, this month, last six months, last twelve months). This will use the query
 ```
SELECT *
FROM cluster_training JOIN tranning
ON cluster_training.training_id=tranning.id
WHERE cluster_training.id = :cluster_id
  AND training.date IN :timerange 
ORDER by cluster_training.pace
LIMIT 10
```
Unfortunately, this query will force the DB to iterate the BTree over the best activities by pace, until at least ten activities that match the date constraint are found.

**Activity-view.**
The activity view is extended with a new section, displaying the cluster matching the activity. Suppose that we have a running activity of 6 km. According with the default running cluster, this activity matches two clusters: >= 1km and >= 5km. In this case, the record section will have two tabs, one for each cluster. Let also assume that this activity is the 100th best activity on the first cluster. The record section will display (i) the top 4 activities for the cluster, (ii) the position in the record chart of the current activity in the cluster, (iii) the worst two activities preceding the current one and (iv) the best activities following the current one. Namely:
```
1 a1 v2
2 a2 v2
3 a3 v3
4 a4 v4
……...
98 a98 v98
99 a99 v99
100 a100 v100
101 a101 v101
102 a102 v102
```
These information can be retrieved with the following queries:
(ii) position of the current activity in the cluster
```
SELECT count(*)
FROM cluster_training
WHERE cluster_training.id = :cluster_id
  AND (cluster_training.pace > :pace
  OR (cluster_training.pace = :pace 
  AMD cluster_training.training_id < cluster_training.training_id))
```
(i) top 4 activities
```
SELECT *
FROM cluster_training JOIN tranning
ON cluster_training.training_id=tranning.id
WHERE cluster_training.id = :cluster_id
ORDER by cluster_training.pace
LIMIT 4
```
(iii) the worst two activities preceding the current one 
```
SELECT *
FROM cluster_training JOIN tranning
ON cluster_training.training_id=tranning.id
WHERE cluster_training.id = :cluster_id
   AND (cluster_training.pace > :pace
  OR (cluster_training.pace = :pace 
  AMN cluster_training.training_id < cluster_training.training_id))
ORDER BY cluster_training.pace DESC
LIMIT 2
```
(iv) the best two activities following the current one 
```
SELECT *
FROM cluster_training JOIN tranning
ON cluster_training.training_id=tranning.id
WHERE cluster_training.id = :cluster_id
   AND NOT (cluster_training.pace > :pace
   OR (cluster_training.pace = :pace 
   AMD cluster_training.training_id < cluster_training.training_id))
ORDER BY cluster_training.pace
LIMIT 2
```
Notice that none of these queries require a table scan, but can be solved by only accessing the index.
An additional functionality can be display the records by time range (e.g. all, this year, last year, this month, last six months, last twelve months, activity year, activity month, up to the activity date, etc). As discussed in for the record plugin, in this case the queries will force the DB to iterate the BTree over the best activities by pace, until the activities that match the date constraint are found.

**Maintaining the cluster_training table**
The cluster_training table has to be updated when something is updated in the system. Clearly, the rows have to be dropped if the corresponding sport, activity or cluster are deleted. Moreover, whenever a cluster is created, the system should query all activities that match the new cluster and add a new row for every one. Similarly, if the cluster parameter are modified (i.e. the distance threshold), the corresponding cluster_training rows should be deleted and the new matching activities re-inserted.
The same actions must be performed for activities. If an activity is created, a row must be added for every cluster that matches the activity. If an activity is updated, the corresponding rows of cluster_training should be deleted and a new row should be added for every matching cluster.

**Design**
According with the current design, we can not implement the new record system as a stand-alone plug-in. The record plug-in can be provided as a stand alone module. However:
we need to add a new section to the activity view. This is currently not possible via plug-in
we need to add a configuration section. The plugin configuration is probably not enough
we need to perform maintenance operation on some events (creation/modification of activities, creation of sport).
My guess is that all maintenance operations and queries can be executed efficiently, with the exception of creation/modification of clusters. For this reason we can imagine to delegate this task to a scheduled task that is executed in background.

**Additional notes**
The new record mechanism can possibly make deprecated a part of the race/altitude plug-ins.

**Comment by laufhannes** [comment]
Thanks for all these details!

**Clusters**
I like the idea of configurable clusters. Some standard distance thresholds are a good set of default clusters for running. But as clusters may not always depend on a distance (e.g. Cooper test, i.e. exactly 12 minutes), additional columns have to be more complex than just a single column for a distance threshold. A set of triples for `field`, `operator` and `value` sounds good, but I'm not sure what's the best database structure for that.

**cluster_training table**
I see the need for these values in terms of performance, but it's not trivial to keep this table consistent with `runalyze_training`. Do we really need more than pace (or any other field of interest, but still, only one field)? If we start to add vdot, we probably add a lot of other fields later on. Maybe it's better to concentrate on one specific field of interest for each cluster which defines the ordering.

**Maintaining the cluster_training table**
I'm not sure how heavy this task is for a) local installations and b) our online version. We definitely need a way to handle this without hitting any server limits. (I don't know much about 'workers' and they are probably not that easy to realize.)

**Design**
Yes, that's definitely a core feature - but that's absolutely okay.

---

A short clarification that came into my mind: All this stuff we are discussing here is about 'total' values of activities, right? Any activity of 10k in 4:00/km will be inserted with 4:00/km into the 1k-cluster, right? All this is not about 'best segments', i.e. trying to find the best kilometer within these 10k and save that one?

---

**Todos**
 - [ ] define an exact database structure for clusters
 - [ ] define an exact database structure for cluster_training

**Comment by mipapo** [comment]
+1 via Ticket #001037

**Comment by laufhannes** [comment]
Depends on #1989 for efficient recalculations.
